### PR TITLE
clean up PeriodIndex constructor

### LINKED ISSUE #13232: PeriodIndex with float input inconsistency
#### Code Sample, a copy-pastable example if possible

```
pd.PeriodIndex([1.1], freq='M')
# ValueError: Given date string not likely a datetime.

pd.PeriodIndex(np.array([1.1]), freq='M')
# PeriodIndex(['1970-02'], dtype='int64', freq='M'
```
#### Expected Output

Consistently raise or coerce to `Period`
#### output of `pd.show_versions()`

0.18.1

CC: @MaximilianR

### PR DESCRIPTION
**Pull request by max-sixty**
- [x] closes #13232 
- [x] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry

Material clean up of PeriodIndex constructor, which was doing a few weird things (https://github.com/pydata/pandas/issues/13232#issuecomment-220788816), and generally getting messy.

### CODE REVIEW

File: pandas/tseries/period.py [hunk 1]
```diff
@@ -41,10 +41,13 @@ def f(self):
 
 def _get_ordinals(data, freq):
     f = lambda x: Period(x, freq=freq).ordinal
+    if len(data) == 0:
+        return np.array([], dtype='int64')
```
Comment (jreback) [inline pandas/tseries/period.py]: you can do if/elsif here (really more a stylistic comment)

File: pandas/tseries/period.py [hunk 2]
```diff
@@ -41,10 +41,13 @@ def f(self):
 
 def _get_ordinals(data, freq):
     f = lambda x: Period(x, freq=freq).ordinal
```
Comment (jreback) [inline pandas/tseries/period.py]: move this to the end of the function

File: pandas/tseries/period.py [hunk 3]
```diff
@@ -211,73 +214,53 @@ def _generate_range(cls, start, end, periods, freq, fields):
 
     @classmethod
     def _from_arraylike(cls, data, freq, tz):
```
Comment (jreback) [inline pandas/tseries/period.py]: I would prob move this entire function into ``__new__``, then its all in 1 place (I don't think its called anywhere else)

**Comment by MaximilianR** [comment]
@jreback One of the reasons the `PeriodIndex` constructer is so convoluted is there are some odd cases supported (I imagine through many bugfixes & iterations!). One of them is a _list of ints representing years_. An array of ints are interpreted as ordinals, but a list is coerced to years within the `Period` constructor.

Here's an example: https://github.com/SixtyCapital/pandas/blob/master/pandas/tseries/tests/test_period.py#L2216

What do you think about not allowing that?

**Comment by jreback** [comment]
@MaximilianR I think that's convenient. The issue is you need a constructor that only deals with codes (otherwise its ambiguous), ok; let's rename ``_from_arraylike`` to ``from_codes`` and make it public; that will solve the ambiguity problem then (and also provide the internal mechanism for creation which is THE way to create from codes).

**Comment by MaximilianR** [comment]
OK, but then you have these being different results: 
```
pd.PeriodIndex([2007, 2008], freq='A')  # years
pd.PeriodIndex(np.array([2007, 2008]), freq='A')  # ordinals 
pd.PeriodIndex(np.array([2007, 2008], dtype='object'), freq='A')  # ?
```

Agree on separation of concerns there - I have `._simple_new` handling ordinals, and `__new__` handling the other cases now.

But we still need `pd.PeriodIndex(idx.values, idx.freq)` to treat the values as ordinals rather than years (otherwise would break a fundamental tenet of indexes). So we're left with the above. 
I can add back the branching logic for lists vs. arrays if you really think it's a helpful use case?

**Comment by jreback** [comment]
hmm, those should all do the same thing.

**Comment by MaximilianR** [comment]
> hmm, those should all do the same thing.

Right. So we can either allow supplying years as ints OR those all the same thing: choose one!
Or a creative solution?

**Comment by jreback** [comment]
actually looking closer I don't think integers are EVER interpreted as ordinals directly, instead the integers are intrepreted according to the frequency. Can you give a counter-example?

**Comment by MaximilianR** [comment]
> actually looking closer I don't think integers are EVER interpreted as ordinals directly, 

From some test failures earlier, but without much exploration on its frequency): in some indexing operations, `PeriodIndex(some_values_from_a_periodindex)` is called. 
There may be ways around that *but* are you sure we want to violate the `idx._constructor(idx.values) == idx` consistency?

> instead the integers are intrepreted according to the frequency.

You need the freq to convert an ordinal to a Period, for sure

**Comment by jreback** [comment]
> ``idx._constructor(idx.values) == idx``
``idx.values`` is an implementation detail. I have no problem with forcing US (meaning you!) to change internal access methods to ``PeriodIndex.from_ordinals(idx.values)`` if you want to construct from ordinals and leave ``PeriodIndex([2000,2001],freq='A-JUNE')`` to work as the public contstructor.

IOW separating out ordinal construction.

**Comment by MaximilianR** [comment]
OK, let me reflect a bit. 

My current thinking depends what you mean by 'internal access methods` - absolutely fine if that's _within_ the PeriodIndex machinery. My fear is that there's a lot of pandas that relies on assertion above, and that going down this road would make `PeriodIndex` need _more_ special casing throughout the library. And I would love to see less, both for `PeriodIndex` and everything else!

One option would be to move towards the Index interface relying on `._simple_new` to interpret any output of the values of an Index. I think that probably isn't a big change.

**Comment by jreback** [comment]
> My current thinking depends what you mean by 'internal access methods- absolutely fine if that's _within_ the PeriodIndex machinery. My fear is that there's a lot of pandas that relies on assertion above, and that going down this road would makePeriodIndexneed _more_ special casing throughout the library. And I would love to see less, both forPeriodIndex` and everything else!


You mean that people are actually constructing this like ``PeriodIndex(idx.values)``? and expecting it to work? in external code, really? 

If you can fix up the internal consistency, does anything actually break? (e.g. create ``PeriodIndex.from_ordinals`` then move all internal code to use that).

**Comment by MaximilianR** [comment]
> You mean that people are actually constructing this like PeriodIndex(idx.values)? and expecting it to work? in external code, really?

No! But there are places in pandas this happens outside of the PeriodIndex class. I need to turn this off and see what breaks - I suspect a reasonable amount but not a torrent

File: pandas/tseries/period.py [hunk 4]
```diff
@@ -211,79 +247,16 @@ def _generate_range(cls, start, end, periods, freq, fields):
...
 
     @classmethod
-    def _from_arraylike(cls, data, freq, tz):
-        if not isinstance(data, (np.ndarray, PeriodIndex,
-                                 DatetimeIndex, Int64Index)):
-            if lib.isscalar(data) or isinstance(data, Period):
-                raise ValueError('PeriodIndex() must be called with a '
-                                 'collection of some kind, %s was passed'
-                                 % repr(data))
+    def _simple_new(cls, data, name=None, freq=None, **kwargs):
```
Comment (MaximilianR) [inline pandas/tseries/period.py]: This needs to be called `data` rather than `values` in order for the current implementation of pickling to work

Comment (MaximilianR) [inline pandas/tseries/period.py]: More generally: I haven't got my head around why we need all the custom pickling infrastructure around pickling with indexes - isn't reconstructing the object's `__dict__` enough?

Comment (jreback) [inline pandas/tseries/period.py]: there *isn't* any current infrastructure, ``Index.__reduce__/__setstate__`` does everything generally. Except for the back-compat stuff (which at some point should just take out). Everythings done as a dict *now* (was a tuple a few versions ago).

Comment (MaximilianR) [inline pandas/tseries/period.py]: By 'infrastructure' I mean the `Index.__reduce__/__setstate__` methods?

Comment (MaximilianR) [inline pandas/tseries/period.py]: By 'infrastructure' I mean the `Index.__reduce__/__setstate__` methods themselves. 
Or you're saying they're only there to allow back compat? And that we could remove those, default to the normal python machinery, and it would work the same for everything from recent versions?

Comment (jreback) [inline pandas/tseries/period.py]: no I don't think it would work. You have to tell pickle what to do. E.g. we want to pickle only *certain* attributes and the data, nothing more. I don't think it works correctly otherwise (it tries to do too much).

You can try removing them but (aside from back-compat) I don't think it would work, but I could be wrong.

In any event I don't see it as a big deal; you want custom actions and that's what the special methods are for.

otherwise its *too* presumptous.

Comment (MaximilianR) [inline pandas/tseries/period.py]: Had a quick go at simplifying it - this passes and takes out all of the custom stuff except for back-compat + a one line `__getstate__`. But also agree, not a big deal
https://github.com/SixtyCapital/pandas/compare/period-float...SixtyCapital:use-getstate

File: pandas/core/algorithms.py [hunk 5]
```diff
@@ -304,17 +304,18 @@ def _value_counts_arraylike(values, dropna=True):
...
     orig = values
 
+    if is_period:
+        from pandas.tseries.period import PeriodIndex
+        index = PeriodIndex(values)
+        freq = index.freq
+        values = index.values
+
     from pandas.core.series import Series
     values = Series(values).values
```
Comment (jreback) [inline pandas/core/algorithms.py]: maybe reorg this a bit, we should catch the datetime/timedelta/period ones before we do these coercion lines

File: pandas/indexes/base.py [hunk 6]
```diff
@@ -67,6 +67,8 @@ def _new_Index(cls, d):
     """ This is called upon unpickling, rather than the default which doesn't
     have arguments and breaks __new__
     """
+    if issubclass(cls, ABCPeriodIndex):
```
Comment (jreback) [inline pandas/indexes/base.py]: why can't you just call ``__new__`` here like verything else? that's sort of the point, we are trying to eliminate special cases.

File: pandas/tests/indexes/test_datetimelike.py [hunk 7]
```diff
@@ -867,7 +867,7 @@ def test_repeat(self):
         # GH10183
         idx = pd.period_range('2000-01-01', periods=3, freq='D')
         res = idx.repeat(3)
-        exp = PeriodIndex(idx.values.repeat(3), freq='D')
+        exp = PeriodIndex._simple_new(idx.values.repeat(3), freq='D')
```
Comment (jreback) [inline pandas/tests/indexes/test_datetimelike.py]: this is why we need a ``.from_ordials``. We really do not want to have to use ``_simple_new`` except internally (to the Index classes). It NOT public at all.

File: pandas/indexes/base.py (hunk 6 above)
Comment (MaximilianR) [inline pandas/indexes/base.py]: Because `__new__` no longer takes ordinals (wasn't that the point of this 😄 )

File: pandas/tseries/period.py [hunk 8]
```diff
@@ -211,96 +247,23 @@ def _generate_range(cls, start, end, periods, freq, fields):
         return subarr, freq
 
     @classmethod
-    def _from_arraylike(cls, data, freq, tz):
```
Comment (jreback) [inline pandas/tseries/period.py]: I am ok with a ``.from_ordinals``

**Comment by jreback** [comment]
@MaximilianR ok does look a lot cleaner

File: pandas/tests/indexes/test_datetimelike.py (hunk 7 above)
Comment (MaximilianR) [inline pandas/tests/indexes/test_datetimelike.py]: OK - I can have `.from_ordinals` point directly to `._simple_new`, or were you thinking something else?

Comment (jreback) [inline pandas/tests/indexes/test_datetimelike.py]: it *could* do simple new if that works (though you also need some validation, it can ONLY be an integer array/list at that point).

File: pandas/indexes/base.py (hunk 6 above)
Comment (MaximilianR) [inline pandas/indexes/base.py]: This change would do away with this stuff: https://github.com/SixtyCapital/pandas/compare/period-float...SixtyCapital:use-getstate

Comment (jreback) [inline pandas/indexes/base.py]: I would rather just make this call PeriodIndex.from_ordinals then

**Comment by MaximilianR** [comment]
I've been remiss about the final fixes here. Will need a couple of weeks on other commitments but I will get to this

Unless @sinhrks is going to spoil all my hard work with moving PeriodIndex to be a proper type

**Comment by sinhrks** [comment]
I'm using ``PeriodIndex`` as internal of ``PeriodDtype`` (like ``DatetimeTZ``), the improvement of ``PeriodIndex`` should not be spoiled:)

**Comment by jorisvandenbossche** [comment]
@MaximilianR do you have time to update this? (the period dtype for index PR has been merged) The we could get it in 0.19

**Comment by MaximilianR** [comment]
I'll have a look in the next couple of weeks. I fear the underlying may have changed enough that this is no longer valid / needed, but let me have a look and assess

**Comment by jreback** [comment]
@MaximilianR can you update

**Comment by MaximilianR** [comment]
Yes, apologies for my slowness with this + a couple of others. I'll have a plug over / after Thanksgiving. Thanks for the ping

**Comment by jreback** [comment]
@MaximilianR can you rebase

**Comment by MaximilianR** [comment]
Just tried rebasing. Painful. Good lesson for getting smaller changes in faster. Will attempt to resolve

**Comment by MaximilianR** [comment]
@jreback green!
🤸‍♀️

File: doc/source/whatsnew/v0.19.2.txt [hunk 9]
```diff
@@ -75,7 +75,7 @@ Bug Fixes
 - Bug in resampling a ``DatetimeIndex`` in local TZ, covering a DST change, which would raise ``AmbiguousTimeError`` (:issue:`14682`)
 
 - Bug in indexing that transformed ``RecursionError`` into ``KeyError`` or ``IndexingError`` (:issue:`14554`)
-
+- Cleaned up ``PeriodIndex`` constructor, including raising on floats more consistently (:issue:`13277`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.19.2.txt]: move to 0.20.0

File: pandas/indexes/base.py [hunk 10]
```diff
@@ -86,6 +86,11 @@ def _new_Index(cls, d):
     """ This is called upon unpickling, rather than the default which doesn't
     have arguments and breaks __new__
     """
+    from pandas.types.generic import ABCPeriodIndex
```
Comment (jreback) [inline pandas/indexes/base.py]: this can be imported at the very top (thats the point of the generics)

Comment (jreback) [inline pandas/indexes/base.py]: is this for backward compat?
if so pls add a comment.

File: pandas/tseries/period.py [hunk 11]
```diff
@@ -308,13 +280,18 @@ def _simple_new(cls, values, name=None, freq=None, **kwargs):
             else:
                 return cls(values, name=name, freq=freq, **kwargs)
 
+        return cls._from_ordinals(values, name, freq, **kwargs)
+
+    @classmethod
+    def _from_ordinals(cls, values, name=None, freq=None, **kwargs):
```
Comment (jreback) [inline pandas/tseries/period.py]: I see that this is private, ok with that. which makes sense as user's don't need to know how the sausage is made.

File: pandas/indexes/base.py (hunk 10 above)
Comment (MaximilianR) [inline pandas/indexes/base.py]: Yeah this is a bit of legacy mess. Could do with a clean up. We'll always need `_new_Index` for existing pickles, but with some basic clean up we could remove a bunch of code and this would be redundant for future pickles

File: pandas/indexes/base.py [hunk 12]
```diff
@@ -86,6 +86,12 @@ def _new_Index(cls, d):
     """ This is called upon unpickling, rather than the default which doesn't
     have arguments and breaks __new__
     """
+    # required for backward compat, because PI can't be instantiated with
+    # ordinals through __new__ GH #13277
+    if issubclass(cls, ABCPeriodIndex):
+        if d['data'].dtype == 'int64':
+            values = d.pop('data')
```
Comment (jreback) [inline pandas/indexes/base.py]: I think its worth putting this in place (but obviously for PeriodIndex) in ``__reduce__``, so that current/future pickling will directly call ``_new_PeriodIndex``, e.g.: https://github.com/pandas-dev/pandas/blob/master/pandas/tseries/index.py#L144

I think you *still* will need this, but you can then simplify to:

```
if issubclass(cls, 'ABCPeriodIndex):
    from pandas.tseries.period import _new_PeriodIndex
   return _new_PeriodIndex(cls, **d)
return cls.__new__(cls, **d)
```

which moves the actual logic to the period.py (where it belongs).

also add a TODO to remove this in a future version (from Index).

File: pandas/tseries/period.py [hunk 13]
```diff
@@ -211,11 +210,46 @@ def __new__(cls, data=None, ordinal=None, freq=None, start=None, end=None,
             else:
                 data, freq = cls._generate_range(start, end, periods,
                                                  freq, kwargs)
-        else:
-            ordinal, freq = cls._from_arraylike(data, freq, tz)
-            data = np.array(ordinal, dtype=np.int64, copy=copy)
+            return cls._from_ordinals(data, name=name, freq=freq)
+
+        if isinstance(data, PeriodIndex):
```
Comment (jreback) [inline pandas/tseries/period.py]: can you put some comments on what these various cases cover.

File: pandas/tseries/period.py [hunk 14]
```diff
@@ -211,11 +210,46 @@ def __new__(cls, data=None, ordinal=None, freq=None, start=None, end=None,
             else:
                 data, freq = cls._generate_range(start, end, periods,
                                                  freq, kwargs)
-        else:
```
Comment (jreback) [inline pandas/tseries/period.py]: can you add under ``_simple_new`` and ``_from_ordinals`` pointers to each other and an expl of when to use which for construction (or can do in a block above them both).

File: pandas/tseries/period.py [hunk 15]
```diff
@@ -211,11 +210,46 @@ def __new__(cls, data=None, ordinal=None, freq=None, start=None, end=None,
...
+        if np.issubdtype(data.dtype, np.datetime64):
+            data = dt64arr_to_periodarr(data, freq, tz)
+            return cls._from_ordinals(data, name=name, freq=freq)
+
+        inferred_dtype = infer_dtype(data)
+
+        if inferred_dtype == 'floating' and len(data) > 0:
+            raise TypeError("PeriodIndex can't take floats")
+
+        else:
```
Comment (jreback) [inline pandas/tseries/period.py]: no else here (just fall thru)

File: pandas/tseries/period.py [hunk 16]
```diff
@@ -940,7 +916,8 @@ def _wrap_union_result(self, other, result):
 
     def _apply_meta(self, rawarr):
         if not isinstance(rawarr, PeriodIndex):
-            rawarr = PeriodIndex(rawarr, freq=self.freq)
+            rawarr = PeriodIndex._from_ordinals(rawarr, freq=self.freq,
```
Comment (jreback) [inline pandas/tseries/period.py]: test to hit this?

File: doc/source/whatsnew/v0.20.0.txt [hunk 17]
```diff
@@ -283,6 +283,7 @@ Bug Fixes
 - Bug in ``astype()`` where ``inf`` values were incorrectly converted to integers. Now raises error now with ``astype()`` for Series and DataFrames (:issue:`14265`)
 - Bug in ``DataFrame(..).apply(to_numeric)`` when values are of type decimal.Decimal. (:issue:`14827`)
 - Bug in ``describe()`` when passing a numpy array which does not contain the median to the ``percentiles`` keyword argument (:issue:`14908`)
+- Cleaned up ``PeriodIndex`` constructor, including raising on floats more consistently (:issue:`13277`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: move to api changes section.

**Comment by jreback** [comment]
@MaximilianR looks pretty good. just a change about the unpickling constructor.

File: doc/source/whatsnew/v0.20.0.txt (hunk 17 above)
Comment (jorisvandenbossche) [inline doc/source/whatsnew/v0.20.0.txt]: *if* we put it in the api section, there should be a clear 'change' documented for users to be aware of. Is the raising on floats not more a bug fix?

File: setup.cfg [hunk 18]
```diff
@@ -13,6 +13,7 @@ parentdir_prefix = pandas-
 
 [flake8]
 ignore = E731
+max-line-length = 79
```
Comment (jorisvandenbossche) [inline setup.cfg]: Isn't this the default?

Comment (MaximilianR) [inline setup.cfg]: Some people will have local defaults that are different. This will override those

File: pandas/indexes/base.py (hunk 12 above)
Comment (MaximilianR) [inline pandas/indexes/base.py]: This needs an overall simplification, I'll add a todo. I think that proposal is better but not optimal.
Let's do that in another PR though, or I'll be doing another painful rebase in 2017!

File: pandas/tseries/period.py (hunk 16 above)
Comment (MaximilianR) [inline pandas/tseries/period.py]: Yes I only saw it because it broke a test. But not really sure whether we need the function

File: doc/source/whatsnew/v0.20.0.txt (hunk 17 above)
Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: yes its a bug fix, but maybe highlite it a bit (also could just expand the text a bit)

File: pandas/indexes/base.py (hunk 12 above)
Comment (jreback) [inline pandas/indexes/base.py]: please move now.

**Comment by jorisvandenbossche** [comment]
@MaximilianR Can you run some of the asv benchmarks to check there is no impact on performance? (also the plotting benchmarks, we have see regressions there in the past due to changes in period index)

File: pandas/indexes/base.py (hunk 12 above)
Comment (MaximilianR) [inline pandas/indexes/base.py]: Why do we use `__reduce__` for Indexes rather than `__getstate__`, like we do in the rest of the library? 
I think the pickling code for Indexes could be _much_ simpler. More than happy to do this in a follow up PR. I don't want to get into the habit of blindly making changes I don't see the logic in though

Comment (jreback) [inline pandas/indexes/base.py]: I think this just a historical reason, originall ``Index`` was a sub-class of ``ndarray``, which used ``__reduce__``. IIRC I changed ``Series`` a while back (these use ``__getstate__/__setstate__``, defined in ``pandas.core.generic``, with some helpers for compat.

So short answer is yes, we should probably change this. But for *this* PR, let's not, pls open a new issue though for this pickle change.

Comment (MaximilianR) [inline pandas/indexes/base.py]: OK, appreciate the response.

Happy to put a new method in `period.py`. To be clear though, you want a new method (that we'd have to support for pickle back-compat) even though we have an immediate plan to deprecate it?

Comment (jreback) [inline pandas/indexes/base.py]: it won't be deprecated just won't be needed

**Comment by MaximilianR** [comment]
@jorisvandenbossche Good call. There is one significant slow down, by about 30x: [`time_from_pydatetime`](https://github.com/SixtyCapital/pandas/blob/period-float/asv_bench/benchmarks/period.py#L15), which cooerces python datetimes to Periods.

I've tried to debug this, but am struggling a bit and could do with some help (and general guidance on how to approach these problems). So far:
- The `Period` constructor is being called by each object within `period.extract_ordinals`, and that's what's taking all the time
- The `Period` constructor then calls [`get_period_ordinal`](https://github.com/pandas-dev/pandas/blob/master/pandas/src/period_helper.c#L1151), which somewhere calls into `to_offset`:
```python
 0.183 __new__  pandas/tseries/period.py:183
├─ 0.168 wrapper  pandas/util/decorators.py:65
│  └─ 0.150 to_offset  pandas/tseries/frequencies.py:453
│     ├─ 0.089 get_stride_from_decimal  pandas/tseries/frequencies.py:186
│     │  └─ 0.086 isclose  numpy/core/numeric.py:2375
│     │     ├─ 0.032 within_tol  numpy/core/numeric.py:2434
│     │     └─ 0.021 all  numpy/core/fromnumeric.py:1987
│     │        ├─ 0.013 _all  numpy/core/_methods.py:40
│     │        └─ 0.003 asanyarray  numpy/core/numeric.py:484
│     ├─ 0.019 get_offset  pandas/tseries/frequencies.py:605
│     │  └─ 0.014 copy  pandas/tseries/offsets.py:306
│     │     └─ 0.013 __init__  pandas/tseries/offsets.py:192
│     ├─ 0.014 __mul__  pandas/tseries/offsets.py:412
│     │  └─ 0.006 __init__  pandas/tseries/offsets.py:192
│     │     └─ 0.004 _determine_offset  pandas/tseries/offsets.py:198
│     │        └─ 0.003 <genexpr>  pandas/tseries/offsets.py:203
│     └─ 0.008 split  re.py:195
└─ 0.014 get_freq_code  pandas/tseries/frequencies.py:300
   └─ 0.003 _period_str_to_code  pandas/tseries/frequencies.py:732
```

Does this matter? Coercing 

CC @sinhrks

**Comment by MaximilianR** [comment]
I think I figured it out - `Period._maybe_convert_freq` needs to be called early, otherwise it the pyx functions will take a slow route to get the appropriate freq. Running ASV to confirm now

**Comment by MaximilianR** [comment]
Performance issues solved! I think all all asv changes are from external compute interference given none are period related. Here are all <50% or >200% for transparency:

```
-  112.12ms    54.32ms      0.48  sparse.sparse_frame_constructor.time_sparse_frame_constructor
-    7.34ms     3.48ms      0.47  reshape.reshape_stack_simple.time_reshape_stack_simple
-    5.97ms     2.70ms      0.45  reshape.melt_dataframe.time_melt_dataframe
-    1.00ms   448.65μs      0.45  reindex.FillMethod.time_pad
-    1.87ms   825.54μs      0.44  period.Algorithms.time_value_counts_pindex
-    1.15ms   485.69μs      0.42  reindex.Reindexing.time_reindex_dates
-  429.95μs   173.53μs      0.40  series_methods.series_constructor_no_data_datetime_index.time_series_constructor_no_data_datetime_index
-   24.72ms     4.49ms      0.18  reindex.Duplicates.time_frame_drop_dups_na_inplace
-   18.22ms     2.74ms      0.15  reindex.Reindexing.time_reindex_multiindex

---

+    6.13ms    14.80ms      2.41  hdfstore_bench.HDF5.time_read_store_mixed
+   26.90ms    60.93ms      2.27  groupby.GroupBySuite.time_rank('int', 100)
+    1.48ms     3.23ms      2.19  frame_methods.Interpolate.time_interpolate_some_good
+  537.00μs     1.16ms      2.16  strings.StringEncode.time_encode_decode
+    2.62ms     5.43ms      2.07  strings.StringMethods.time_contains_few_noregex
```

**Comment by MaximilianR** [comment]
@jreback green!

File: pandas/tseries/period.py [hunk 19]
```diff
@@ -205,17 +211,58 @@ def __new__(cls, data=None, ordinal=None, freq=None, start=None, end=None,
...
+                data = period.period_asfreq_arr(data._values,
+                                                base1, base2, 1)
+            return cls._simple_new(data, name=name, freq=freq)
 
-        return cls._simple_new(data, name=name, freq=freq)
+        # not array / index
+        if not isinstance(data, (np.ndarray, PeriodIndex,
+                                 DatetimeIndex, Int64Index)):
+            if is_scalar(data) or isinstance(data, Period):
+                raise ValueError('PeriodIndex() must be called with a '
```
Comment (jreback) [inline pandas/tseries/period.py]: this is generally a TypeError, in fact changing this to ``cls._scalar_data_error(data)`` is appropriate (I am changing on merge)

Comment (jreback) [inline pandas/tseries/period.py]: actually why don't you make this change (as I have a couple of more comments).

```
diff --git a/pandas/tseries/period.py b/pandas/tseries/period.py
index 903b4e3..cefe1cf 100644
--- a/pandas/tseries/period.py
+++ b/pandas/tseries/period.py
@@ -239,9 +239,7 @@ class PeriodIndex(DatelikeOps, DatetimeIndexOpsMixin, Int64Index):
         if not isinstance(data, (np.ndarray, PeriodIndex,
                                  DatetimeIndex, Int64Index)):
             if is_scalar(data) or isinstance(data, Period):
-                raise ValueError('PeriodIndex() must be called with a '
-                                 'collection of some kind, %s was passed'
-                                 % repr(data))
+                cls._scalar_data_error(data)
 
             # other iterable of some kind
             if not isinstance(data, (list, tuple)):
diff --git a/pandas/tseries/tests/test_period.py b/pandas/tseries/tests/test_period.py
index c7fd4ae..8ec2fae 100644
--- a/pandas/tseries/tests/test_period.py
+++ b/pandas/tseries/tests/test_period.py
@@ -1753,7 +1753,7 @@ class TestPeriodIndex(tm.TestCase):
 
         self.assertRaises(ValueError, PeriodIndex, idx._values)
         self.assertRaises(ValueError, PeriodIndex, list(idx._values))
-        self.assertRaises(ValueError, PeriodIndex,
+        self.assertRaises(TypeError, PeriodIndex,
                           data=Period('2007', freq='A'))
 
         result = PeriodIndex(iter(idx))
```

File: pandas/tseries/period.py [hunk 20]
```diff
@@ -205,17 +211,58 @@ def __new__(cls, data=None, ordinal=None, freq=None, start=None, end=None,
...
+                                 % repr(data))
+
+            # other iterable of some kind
+            if not isinstance(data, (list, tuple)):
+                data = list(data)
+
+            data = np.asarray(data)
+
+        # datetime other than period
+        if np.issubdtype(data.dtype, np.datetime64):
```
Comment (jreback) [inline pandas/tseries/period.py]: this should be ``is_datetime64_dtype``. But not sure what should happend if you have tz-aware or datetime64 (but not ns) passed. We may not have tests for this. You *could* handle datetime64tz by pretty easily though (happy to have this done in another PR as well).

File: pandas/tseries/period.py [hunk 21]
```diff
@@ -237,84 +284,34 @@ def _generate_range(cls, start, end, periods, freq, fields):
...
+        """
+        Values can be any type that can be coerced to Periods.
+        Ordinals in an ndarray are fastpath-ed to `_from_ordinals`
+        """
         if not is_integer_dtype(values):
             values = np.array(values, copy=False)
-            if (len(values) > 0 and is_float_dtype(values)):
+            if len(values) > 0 and is_float_dtype(values):
                 raise TypeError("PeriodIndex can't take floats")
             else:
```
Comment (jreback) [inline pandas/tseries/period.py]: can eliminate the else clause (and just do the return)

File: pandas/tseries/period.py [hunk 22]
```diff
@@ -325,7 +322,6 @@ def _shallow_copy_with_infer(self, values=None, **kwargs):
 
     def _shallow_copy(self, values=None, **kwargs):
         if kwargs.get('freq') is None:
```
Comment (jreback) [inline pandas/tseries/period.py]: I would explicitly list ``freq=None`` I think in the signature (then pass it directly) as its so crucial to the PeriodIndex nice to be noted

**Comment by jreback** [comment]
couple of comments. lgtm otherwise. if you'd fix / create a new issue (for the handling of datetime-aware tz, other datetimes), then ok with merging.

**Comment by jreback** [comment]
can you rebase / update

**Comment by jorisvandenbossche** [comment]
@MaximilianR You have a bunch of errors, all seems related to the use of `np.issubdtype(data.dtype, is_datetime64_dtype)` (not sure what changed here)

**Comment by jreback** [comment]
this lgtm. any perf impacts?

**Comment by jreback** [comment]
thanks @MaximilianR 

figured out to push this in so no rebases are needed!
